File: dotnet-desktop-guide/winforms/migration/snippets/clipboard-dataobject-net10/net/csharp/CustomTypesExamples.cs
using System;
using System.Text.Json.Serialization;
using System.Windows.Forms;

namespace ClipboardExamples
{
    public class CustomTypesExamples
    {
        // <SimpleCustomTypes>
        // Records work without any attributes.
        public record PersonInfo(string Name, int Age, string Email);

        // Simple classes serialize all public properties automatically.
        public class DocumentMetadata
        {
            public string Title { get; set; }
            public DateTime Created { get; set; }
            public string Author { get; set; }
        }

        // Structs with public properties work seamlessly.
        public struct Point3D
        {
            public double X { get; set; }
            public double Y { get; set; }
            public double Z { get; set; }
        }
        // </SimpleCustomTypes>

        // <JsonAttributesExample>
        public class ClipboardFriendlyType
        {
            // Include a field that normally isn't serialized
            [JsonInclude]
            private int _privateData;

            // Public properties are always serialized
            public string Name { get; set; }
            
            // Exclude sensitive or non-essential data
            [JsonIgnore]
            public string InternalId { get; set; }
            
            // Handle property name differences for compatibility
            [JsonPropertyName("display_text")]
            public string DisplayText { get; set; }
            
            // Control null value handling
            [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
            public string OptionalField { get; set; }
        }
        // </JsonAttributesExample>

        // <CustomTypesClipboardOperations>
        public static void CustomTypesClipboardOperationsExample()
        {
            var data = new ClipboardFriendlyType 
            { 
                Name = "Sample", 
                DisplayText = "Sample Display Text",
                InternalId = "internal-123" // This property isn't serialized due to [JsonIgnore]
            };

            Clipboard.SetDataAsJson("MyAppData", data);

            if (Clipboard.TryGetData("MyAppData", out ClipboardFriendlyType retrieved))
            {
                Console.WriteLine($"Retrieved: {retrieved.Name}");
                // retrieved.InternalId is null because of [JsonIgnore]
            }
        }
        // </CustomTypesClipboardOperations>
    }
}

File: dotnet-desktop-guide/winforms/migration/snippets/clipboard-dataobject-net10/net/vb/CustomTypesExamples.vb
Imports System
Imports System.Text.Json.Serialization
Imports System.Windows.Forms

Namespace ClipboardExamples
    Public Class CustomTypesExamples
        ' <SimpleCustomTypes>
        ' Simple classes serialize all public properties automatically.
        Public Class DocumentMetadata
            Public Property Title As String
            Public Property Created As DateTime
            Public Property Author As String
        End Class

        ' Structs with public properties work seamlessly.
        Public Structure Point3D
            Public Property X As Double
            Public Property Y As Double
            Public Property Z As Double
        End Structure
        ' </SimpleCustomTypes>

        ' <JsonAttributesExample>
        Public Class ClipboardFriendlyType
            ' Include a field that normally isn't serialized
            <JsonInclude>
            Private _privateData As Integer

            ' Public properties are always serialized
            Public Property Name As String

            ' Exclude sensitive or non-essential data
            <JsonIgnore>
            Public Property InternalId As String

            ' Handle property name differences for compatibility
            <JsonPropertyName("display_text")>
            Public Property DisplayText As String

            ' Control null value handling
            <JsonIgnore(Condition:=JsonIgnoreCondition.WhenWritingNull)>
            Public Property OptionalField As String
        End Class
        ' </JsonAttributesExample>

        ' <CustomTypesClipboardOperations>
        Public Shared Sub CustomTypesClipboardOperationsExample()
            Dim data As New ClipboardFriendlyType With {
                .Name = "Sample",
                .DisplayText = "Sample Display Text",
                .InternalId = "internal-123" ' This property isn't serialized due to <JsonIgnore>
            }

            Clipboard.SetDataAsJson("MyAppData", data)

            Dim retrieved As ClipboardFriendlyType = Nothing
            If Clipboard.TryGetData("MyAppData", retrieved) Then
                Console.WriteLine($"Retrieved: {retrieved.Name}")
                ' retrieved.InternalId is null because of <JsonIgnore>
            End If
        End Sub
        ' </CustomTypesClipboardOperations>
    End Class
End Namespace
